File: viewpump/src/main/java/io/github/inflationx/viewpump/InflateRequest.java
package io.github.inflationx.viewpump;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.view.View;

public class InflateRequest {
    private final String name;
    private final Context context;
    private final AttributeSet attrs;
    private final View parent;
    private final FallbackViewCreator fallbackViewCreator;

    private InflateRequest(Builder builder) {
        name = builder.name;
        context = builder.context;
        attrs = builder.attrs;
        parent = builder.parent;
        fallbackViewCreator = builder.fallbackViewCreator;
    }

    @NonNull
    public String name() {
        return name;
    }

    @NonNull
    public Context context() {
        return context;
    }

    @Nullable
    public AttributeSet attrs() {
        return attrs;
    }

    @Nullable
    public View parent() {
        return parent;
    }

    @NonNull
    public FallbackViewCreator fallbackViewCreator() {
        return fallbackViewCreator;
    }

    @NonNull
    public static Builder builder() {
        return new Builder();
    }

    @NonNull
    public Builder toBuilder() {
        return new Builder(this);
    }

    @NonNull
    @Override
    public String toString() {
        return "InflateRequest{" +
                "name='" + name + '\'' +
                ", context=" + context +
                ", attrs=" + attrs +
                ", parent=" + parent +
                ", fallbackViewCreator=" + fallbackViewCreator +
                '}';
    }

    public static final class Builder {
        private String name;
        private Context context;
        private AttributeSet attrs;
        private View parent;
        private FallbackViewCreator fallbackViewCreator;

        private Builder() { }

        private Builder(InflateRequest request) {
            this.name = request.name;
            this.context = request.context;
            this.attrs = request.attrs;
            this.parent = request.parent;
            this.fallbackViewCreator = request.fallbackViewCreator;
        }

        public Builder name(@NonNull String name) {
            this.name = name;
            return this;
        }

        public Builder context(@NonNull Context context) {
            this.context = context;
            return this;
        }

        public Builder attrs(@Nullable AttributeSet attrs) {
            this.attrs = attrs;
            return this;
        }

        public Builder parent(@Nullable View parent) {
            this.parent = parent;
            return this;
        }

        public Builder fallbackViewCreator(@NonNull FallbackViewCreator fallbackViewCreator) {
            this.fallbackViewCreator = fallbackViewCreator;
            return this;
        }

        public InflateRequest build() {
            if (name == null) {
                throw new IllegalStateException("name == null");
            }
            if (context == null) {
                throw new IllegalStateException("context == null");
            }
            if (fallbackViewCreator == null) {
                throw new IllegalStateException("fallbackViewCreator == null");
            }
            return new InflateRequest(this);
        }
    }
}

File: viewpump/src/main/java/io/github/inflationx/viewpump/InflateRequest.kt
package io.github.inflationx.viewpump

import android.content.Context
import android.util.AttributeSet
import android.view.View

data class InflateRequest(
    @get:JvmName("name")
    val name: String,
    @get:JvmName("context")
    val context: Context,
    @get:JvmName("attrs")
    val attrs: AttributeSet? = null,
    @get:JvmName("parent")
    val parent: View? = null,
    @get:JvmName("fallbackViewCreator")
    val fallbackViewCreator: FallbackViewCreator
) {

  fun toBuilder(): Builder {
    return Builder(this)
  }

  class Builder {
    private var name: String? = null
    private var context: Context? = null
    private var attrs: AttributeSet? = null
    private var parent: View? = null
    private var fallbackViewCreator: FallbackViewCreator? = null

    internal constructor()

    internal constructor(request: InflateRequest) {
      this.name = request.name
      this.context = request.context
      this.attrs = request.attrs
      this.parent = request.parent
      this.fallbackViewCreator = request.fallbackViewCreator
    }

    fun name(name: String) = apply {
      this.name = name
    }

    fun context(context: Context) = apply {
      this.context = context
    }

    fun attrs(attrs: AttributeSet?) = apply {
      this.attrs = attrs
    }

    fun parent(parent: View?) = apply {
      this.parent = parent
    }

    fun fallbackViewCreator(fallbackViewCreator: FallbackViewCreator) = apply {
      this.fallbackViewCreator = fallbackViewCreator
    }

    fun build() =
        InflateRequest(name = name ?: throw IllegalStateException("name == null"),
            context = context ?: throw IllegalStateException("context == null"),
            attrs = attrs,
            parent = parent,
            fallbackViewCreator = fallbackViewCreator ?: throw IllegalStateException("fallbackViewCreator == null")
        )
  }

  companion object {

    @JvmStatic
    fun builder(): Builder {
      return Builder()
    }
  }
}
